File: frontend/src/types/featureFlag.ts

// ============================================================
// FEATURE FLAG TYPES
// ============================================================

// ============================================================
// USER DETAILS
// ============================================================

export interface FeatureFlagUser {
  id?: number;
  name?: string;
  email?: string;
  role?: string;
  role_id?: number;
}

// ============================================================
// FEATURE FLAG
// ============================================================

export interface FeatureFlag {
  id: number;

  name: string;

  key: string;

  description?: string;

  is_enabled?: boolean;

  is_active?: boolean;

  default_value?: string | boolean | number | null;

  value?: string | boolean | number | null;

  created_by_id?: number;

  created_at?: string;

  updated_at?: string;

  created_by?: FeatureFlagUser;

  updated_by?: FeatureFlagUser;
}

// ============================================================
// CREATE FEATURE FLAG REQUEST
// ============================================================

export interface FeatureFlagCreateRequest {
  name: string;

  key: string;

  description?: string;

  is_enabled?: boolean;

  default_value?: string | boolean | number | null;
}

// ============================================================
// UPDATE FEATURE FLAG REQUEST
// ============================================================

export interface FeatureFlagUpdateRequest {
  name?: string;

  key?: string;

  description?: string;

  default_value?: string | boolean | number | null;

  is_enabled?: boolean;

  is_active?: boolean;
}

// ============================================================
// ENABLE / DISABLE RESPONSE
// ============================================================

export interface FeatureFlagToggleResponse {
  success?: boolean;

  message?: string;

  feature?: FeatureFlag;

  enabled_by?: FeatureFlagUser;

  disabled_by?: FeatureFlagUser;

  data?:
    | FeatureFlag
    | {
        feature?: FeatureFlag;

        enabled_by?: FeatureFlagUser;

        disabled_by?: FeatureFlagUser;
      };
}

// ============================================================
// CREATE RESPONSE
// ============================================================

export interface FeatureFlagCreateResponse {
  success?: boolean;

  message?: string;

  feature?: FeatureFlag;

  created_by?: FeatureFlagUser;

  data?:
    | FeatureFlag
    | {
        feature?: FeatureFlag;

        created_by?: FeatureFlagUser;
      };
}

// ============================================================
// UPDATE RESPONSE
// ============================================================

export interface FeatureFlagUpdateResponse {
  success?: boolean;

  message?: string;

  feature?: FeatureFlag;

  updated_by?: FeatureFlagUser;

  data?:
    | FeatureFlag
    | {
        feature?: FeatureFlag;

        updated_by?: FeatureFlagUser;
      };
}

// ============================================================
// DELETE RESPONSE
// ============================================================

export interface FeatureFlagDeleteResponse {
  success?: boolean;

  message?: string;

  feature_id?: number;

  deleted_by?: FeatureFlagUser;

  data?: {
    feature_id?: number;

    deleted_by?: FeatureFlagUser;
  };
}

// ============================================================
// GET SINGLE FEATURE FLAG RESPONSE
// ============================================================

export interface FeatureFlagSingleResponse {
  success?: boolean;

  message?: string;

  feature?: FeatureFlag;

  retrieved_by?: FeatureFlagUser;

  data?:
    | FeatureFlag
    | {
        feature?: FeatureFlag;

        retrieved_by?: FeatureFlagUser;
      };
}

// ============================================================
// GET FEATURE FLAGS RESPONSE
// ============================================================

export interface FeatureFlagListResponse {
  success?: boolean;

  message?: string;

  total?: number;

  features?: FeatureFlag[];

  retrieved_by?: FeatureFlagUser;

  data?:
    | FeatureFlag[]
    | {
        features?: FeatureFlag[];

        total?: number;

        retrieved_by?: FeatureFlagUser;
      };
}

// ============================================================
// FEATURE EVALUATION REQUEST
// ============================================================

export interface FeatureEvaluationRequest {
  user_id?: number;

  environment_id?: number;

  context?: Record<string, any>;
}

// ============================================================
// FEATURE EVALUATION RESULT
// ============================================================

export interface FeatureEvaluation {
  feature_id?: number;

  feature_name?: string;

  feature_key?: string;

  environment_id?: number;

  environment_name?: string;

  user_id?: number;

  user_name?: string;

  value?: string | boolean | number | null;

  is_enabled?: boolean;

  reason?: string;
}

// ============================================================
// FEATURE EVALUATION RESPONSE
// ============================================================

export interface FeatureEvaluationResponse {
  success?: boolean;

  message?: string;

  evaluated_by?: FeatureFlagUser;

  evaluation?: FeatureEvaluation;

  feature?: FeatureEvaluation;

  data?:
    | FeatureEvaluation
    | {
        evaluation?: FeatureEvaluation;

        feature?: FeatureEvaluation;

        evaluated_by?: FeatureFlagUser;
      };
}

File: backend/app/schemas/feature_flag.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field


# ============================================================
# FEATURE CREATE
# ============================================================

class FeatureCreate(BaseModel):

    name: str = Field(
        ...,
        min_length=2,
        max_length=100
    )

    key: str = Field(
        ...,
        min_length=2,
        max_length=100
    )

    description: str | None = Field(
        default=None,
        max_length=500
    )

    is_enabled: bool = False


# ============================================================
# FEATURE UPDATE
# ============================================================

class FeatureUpdate(BaseModel):

    name: str | None = Field(
        default=None,
        min_length=2,
        max_length=100
    )

    key: str | None = Field(
        default=None,
        min_length=2,
        max_length=100
    )

    description: str | None = Field(
        default=None,
        max_length=500
    )

    is_enabled: bool | None = None


# ============================================================
# ROLE DETAILS
# ============================================================

class FeatureRoleResponse(BaseModel):

    id: int
    name: str
    description: str | None = None
    is_active: bool

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# USER DETAILS
# ============================================================

class FeatureUserResponse(BaseModel):

    id: int
    name: str
    email: EmailStr
    is_active: bool

    role: FeatureRoleResponse

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# CREATED BY
# ============================================================

class FeatureCreatedByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# RETRIEVED BY
# ============================================================

class FeatureRetrievedByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# UPDATED BY
# ============================================================

class FeatureUpdatedByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# DELETED BY
# ============================================================

class FeatureDeletedByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# ENABLED BY
# ============================================================

class FeatureEnabledByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# DISABLED BY
# ============================================================

class FeatureDisabledByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# EVALUATED BY
# ============================================================

class FeatureEvaluatedByResponse(BaseModel):

    user: FeatureUserResponse

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# FEATURE RESPONSE
# ============================================================

class FeatureResponse(BaseModel):

    id: int
    name: str
    key: str
    description: str | None = None
    is_enabled: bool

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


# ============================================================
# CREATE RESPONSE
# ============================================================

class FeatureCreateResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    created_by: FeatureCreatedByResponse


# ============================================================
# SINGLE FEATURE RESPONSE
# ============================================================

class FeatureSingleResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    retrieved_by: FeatureRetrievedByResponse


# ============================================================
# LIST RESPONSE
# ============================================================

class FeatureListResponse(BaseModel):

    success: bool
    message: str
    total: int
    features: list[FeatureResponse]
    retrieved_by: FeatureRetrievedByResponse


# ============================================================
# UPDATE RESPONSE
# ============================================================

class FeatureUpdateResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    updated_by: FeatureUpdatedByResponse


# ============================================================
# DELETE RESPONSE
# ============================================================

class FeatureDeleteResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    deleted_by: FeatureDeletedByResponse


# ============================================================
# ENABLE RESPONSE
# ============================================================

class FeatureEnableResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    enabled_by: FeatureEnabledByResponse


# ============================================================
# DISABLE RESPONSE
# ============================================================

class FeatureDisableResponse(BaseModel):

    success: bool
    message: str
    feature: FeatureResponse
    disabled_by: FeatureDisabledByResponse


# ============================================================
# EVALUATE RESPONSE
# ============================================================

class FeatureEvaluateResponse(BaseModel):

    success: bool
    message: str

    feature: FeatureResponse

    # User for whom feature is evaluated
    user: FeatureUserResponse

    enabled: bool

    # Admin/User who performed evaluation
    evaluated_by: FeatureEvaluatedByResponse